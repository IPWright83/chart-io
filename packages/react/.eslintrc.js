module.exports = {
    env: {
        browser: true,
        es2021: true,
    },
    extends: ["eslint:recommended", "plugin:react/recommended", "plugin:@typescript-eslint/recommended"],
    overrides: [
        {
            files: "*.unit.tsx?",
            rules: {
                "valid-jsdoc": "off",
            },
        },
    ],
    parser: "@typescript-eslint/parser",
    parserOptions: {
        ecmaVersion: "latest",
        sourceType: "module",
    },
    plugins: ["react", "@typescript-eslint"],
    ignorePatterns: ["jest.config.js", ".eslintrc.js", "**/jest/*.js", "testUtils/*.ts", "*.stories.tsx"],
    rules: {
        "@typescript-eslint/no-explicit-any": "off",
        "valid-jsdoc": [
            "warn",
            {
                requireReturnType: false,
                requireParamType: false,
                requireReturn: false,
            },
        ],
        "@typescript-eslint/ban-ts-comment": "warn",
    },
};
