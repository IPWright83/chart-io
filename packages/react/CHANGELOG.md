# @chart-it/react-d3

## 0.24.0

### Minor Changes

- f462652: Allow non-array inputs for Scale/Axis

## 0.23.0

### Minor Changes

- 245192e: Allow non-array inputs for Scale/Axis

## 0.22.0

### Minor Changes

- f8166d8: Added support for Legends

## 0.21.0

### Minor Changes

- 17c4895: Test

## 0.20.0

### Minor Changes

- 0862879: Test

## 0.19.0

### Minor Changes

- bb81e3a: Test

## 0.18.0

### Minor Changes

- 494a204: Test
- 5c9ca11: testing

## 0.17.0

### Minor Changes

- 744f129: Updated Chromatic settings

## 0.16.0

### Minor Changes

- feddec9: TypeScript support

### Patch Changes

- Updated dependencies [feddec9]
  - @chart-it/d3@0.2.0
  - @chart-it/detection@0.5.0
  - @chart-it/types@0.5.0

## 0.15.0

### Minor Changes

- b0d2e5e: - Exposed more options to Axis (tickValues)
  - Fixed pointer-events on Markers causing flicker
  - Introduced an `<AutoScale>` component, which is the old `<Scale>`
  - Introduced some new warnings for protection
  - Added the ability to format tooltip popup strings

## 0.14.0

### Minor Changes

- 3f81ca2: Pick stroke color from background for bars/columns
- 3f81ca2: Resolved issue with flickering tooltips due to incorrect pointer-events

## 0.13.0

### Minor Changes

- 175dd66: Pick stroke color from background for bars/columns
- 175dd66: Resolved issue with flickering tooltips due to incorrect pointer-events

## 0.12.0

### Minor Changes

- a6cc33a: Testing changesets

### Patch Changes

- Updated dependencies [a6cc33a]
  - @chart-it/detection@0.4.0

## 0.11.0

### Minor Changes

- aa0b805: Testing changesets

### Patch Changes

- Updated dependencies [aa0b805]
  - @chart-it/detection@0.3.0

## 0.10.0

### Minor Changes

- 29b2703: Fixing peer dependencies

## 0.9.0

### Minor Changes

- 3254c1e: Vite build

## 0.8.0

### Minor Changes

- bc7e7f3: Trying to fix publishing CI

### Patch Changes

- Updated dependencies [bc7e7f3]
  - @chart-it/detection@0.2.0

## 0.7.0

### Minor Changes

- 71eaab3: Tag Test 2

## 0.6.0

### Minor Changes

- 74010a0: This is a test

## 0.5.0

### Minor Changes

- 4d17cab: Added a Rect component
- 4d17cab: Allowed more marker styling

## 0.4.0

### Minor Changes

- 1c8f120: Fixes issues with linkStores in production builds

## 0.3.0

### Minor Changes

- ccedcd6: Updated Storybook docs

## 0.2.0

### Minor Changes

- 856ea4f: Improved selector performance by introducing reselect

## 0.1.33

### Patch Changes

- c2ccdd1: Test
- Updated dependencies [c2ccdd1]
  - @chart-it/detection@0.1.33
