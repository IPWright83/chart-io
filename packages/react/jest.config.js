module.exports = {
    collectCoverageFrom: ["src/**/*.{js,jsx,ts,tsx}", "!src/**/*.d.ts"],
    coveragePathIgnorePatterns: [
        "stories/*.*",
        "types/*.*",
        "index.js",
        "uuidv4.js",
        "interpolateMultiPath.js",
        "hoc/withTheme/*.*",
        "jest",
        "data",
        "lib/testUtils",
        "storybook",
    ],
    moduleFileExtensions: ["web.js", "js", "web.ts", "ts", "web.tsx", "tsx", "json", "web.jsx", "jsx", "node"],
    moduleNameMapper: {
        "^react-native$": "react-native-web",
        "\\.(css|less)$": "identity-obj-proxy",
        "^d3-array$": "<rootDir>/node_modules/d3-array/dist/d3-array.min.js",
        "^d3-axis$": "<rootDir>/node_modules/d3-axis/dist/d3-axis.min.js",
        "^d3-color$": "<rootDir>/node_modules/d3-color/dist/d3-color.min.js",
        "^d3-ease$": "<rootDir>/node_modules/d3-ease/dist/d3-ease.min.js",
        "^d3-format$": "<rootDir>/node_modules/d3-format/dist/d3-format.min.js",
        "^d3-scale$": "<rootDir>/node_modules/d3-scale/dist/d3-scale.min.js",
        "^d3-selection$": "<rootDir>/node_modules/d3-selection/dist/d3-selection.min.js",
        "^d3-shape$": "<rootDir>/node_modules/d3-shape/dist/d3-shape.min.js",
        "^d3-time-format$": "<rootDir>/node_modules/d3-time-format/dist/d3-time-format.min.js",
        "^d3-timer$": "<rootDir>/node_modules/d3-timer/dist/d3-timer.min.js",
        "^d3-transition$": "<rootDir>/node_modules/d3-transition/dist/d3-transition.min.js",
    },
    resetMocks: true,
    roots: ["<rootDir>/src"],
    setupFiles: ["react-app-polyfill/jsdom"],
    setupFilesAfterEnv: ["<rootDir>/src/jest/setupTests.js"],
    testEnvironment: "jsdom",
    testMatch: ["**/*.unit.js", "**/*.unit.jsx", "**/*.unit.ts", "**/*.unit.tsx"],
    transformIgnorePatterns: [
        "[/\\\\]node_modules[/\\\\].+\\.(js|jsx|mjs|cjs|ts|tsx)$",
        "^.+\\.module\\.(css|sass|scss)$",
    ],
    transform: {
        "^.+\\.(js|jsx|mjs|cjs|ts|tsx)$": "ts-jest",
        "^.+\\.css$": "<rootDir>/src/jest/cssTransform.js",
        "^(?!.*\\.(js|jsx|mjs|cjs|ts|tsx|css|json)$)": "<rootDir>/src/jest/fileTransform.js",
    },
    watchPlugins: ["jest-watch-typeahead/filename", "jest-watch-typeahead/testname"],
};
