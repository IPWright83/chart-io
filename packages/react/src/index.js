export * from "./components";
export * from "./hoc";
export * from "./hooks";
export * from "./store";
export * from "./types";
export * from "./utils";
export * from "./constants";
export * from "./themes";
