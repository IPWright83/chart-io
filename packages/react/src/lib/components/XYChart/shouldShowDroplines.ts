/**
 * Determine if droplines should be shown based on the plots used
 * @param  children     The array of React elements
 * @return              True if droplines should be enabled
 */
export function shouldShowDroplines(children: any): boolean {
    if (!children) {
        return true;
    }

    // Just a single child
    if (children.props && children.type) {
        return true;
    }

    const names = children
        .flat()
        .filter((c) => !!c) // Remove undefined children (e.g. condition ?? <component>)
        .map((child) => child.type.name);

    // Disable vertical drop lines when we have multiple Line or Area
    if (names.includes("Areas") || names.includes("Lines")) {
        return false;
    }

    // Check for manually added multiple Line or Area plots
    return names.filter((n) => n === "Area" || n === "Line").length <= 1;
}
