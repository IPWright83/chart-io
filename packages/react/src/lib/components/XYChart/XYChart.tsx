import React from "react";
import { IMargin } from "@chart-io/types";

import { Chart, IChartProps } from "../Chart";
import { Crosshair } from "../Crosshair";
import { Droplines } from "../Droplines";
import { EventReceiver } from "../EventReceiver";
import { LegendOverlay } from "../LegendOverlay";
import { Markers } from "../Markers";
import { RectangleClipPath } from "../ClipPath";
import { TooltipOverlay } from "../TooltipOverlay";
import { ZoomBrush } from "../ZoomBrush";

import { shouldShowDroplines } from "./shouldShowDroplines";

export interface IXYChartProps extends IChartProps {
    /**
     * The type of zoom brush to show if any
     */
    zoomBrush?: "inline" | "overlay";

    /**
     * The margin to apply around a zoomable brush if present
     * @default { left: 0, top: 10, right: 0, bottom: 10 }
     */
    brushMargin?: IMargin;
}

export function XYChart({ children, ...props }: IXYChartProps) {
    const showDroplines = shouldShowDroplines(children);
    const showCrosshair = !showDroplines;

    return (
        <Chart {...props}>
            <EventReceiver />
            <RectangleClipPath>{children}</RectangleClipPath>
            <Markers />
            {showDroplines && <Droplines />}
            {showCrosshair && <Crosshair />}
            {props.zoomBrush && (
                <ZoomBrush type={props.zoomBrush} margin={props.brushMargin}>
                    {children}
                </ZoomBrush>
            )}
            <TooltipOverlay />
            <LegendOverlay />
        </Chart>
    );
}
