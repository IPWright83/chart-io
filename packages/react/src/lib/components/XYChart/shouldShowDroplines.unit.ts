import React from "react";

import { shouldShowDroplines } from "./shouldShowDroplines";

describe("shouldShowDroplines", () => {
    const mockElement = React.createElement("div");

    it("should return true for a single child", () => {
        expect(shouldShowDroplines(mockElement)).toBe(true);
    });

    it("should return true for children excluding lines/areas", () => {
        expect(shouldShowDroplines([mockElement, undefined, mockElement, mockElement])).toBe(true);
    });

    it("should return true if a single Line plot is included", () => {
        const mockLinePlot = React.createElement("Line");

        expect(shouldShowDroplines([mockElement, mockLinePlot])).toBe(true);
    });

    it("should return false if a multiple Line plots are included", () => {
        const mockLinePlot = React.createElement("Line");

        expect(shouldShowDroplines([mockElement, mockLinePlot, mockLinePlot])).toBe(false);
    });

    it("should return false if a Lines plot is included", () => {
        const mockLinesPlot = React.createElement("Lines");

        expect(shouldShowDroplines([mockElement, mockLinesPlot])).toBe(false);
    });

    it("should return true if single Area plot is included", () => {
        const mockAreaPlot = React.createElement("Area");

        expect(shouldShowDroplines([mockElement, mockAreaPlot])).toBe(true);
    });

    it("should return false if multiple Area plots are included", () => {
        const mockAreaPlot = React.createElement("Area");

        expect(shouldShowDroplines([mockElement, mockAreaPlot, mockAreaPlot])).toBe(false);
    });

    it("should return false if an Areas plot is included", () => {
        const mockAreasPlot = React.createElement("Areas");

        expect(shouldShowDroplines([mockElement, mockAreasPlot])).toBe(false);
    });
});
