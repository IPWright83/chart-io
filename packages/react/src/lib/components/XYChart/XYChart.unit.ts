import React from "react";
import { render } from "@testing-library/react";

import { XYChart } from "./XYChart";

describe("XYChart", () => {
    it("should create empty template correctly", () => {
        const { asFragment } = render(<XYChart width={500} height={500} />);

        expect(asFragment()).toMatchSnapshot();
    });
});
