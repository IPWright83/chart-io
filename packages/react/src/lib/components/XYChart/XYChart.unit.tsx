import React from "react";
import { render } from "@testing-library/react";

import { XYChart } from "./XYChart";

jest.mock("../Chart/generateRandomID", () => {
    return { generateRandomID: () => "random-id" };
});

describe("XYChart", () => {
    it("should create empty template correctly", () => {
        const { asFragment } = render(<XYChart width={500} height={500} />);

        expect(asFragment()).toMatchSnapshot();
    });
});
