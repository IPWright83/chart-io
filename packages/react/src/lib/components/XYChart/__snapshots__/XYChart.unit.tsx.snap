// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`XYChart should create empty template correctly 1`] = `
<DocumentFragment>
  <svg
    class="chart-svg"
    height="500"
    style="background-color: rgb(255, 255, 255);"
    width="500"
  >
    <g
      class="g-event-receiver"
    >
      <rect
        class="chart-io event-receiver"
        height="440"
        style="fill: none; pointer-events: all;"
        transform="translate(30, 30)"
        width="440"
      />
    </g>
    <defs>
      <clippath
        id="clip-path-random-id"
      >
        <rect
          height="440"
          width="440"
          x="30"
          y="30"
        />
      </clippath>
    </defs>
    <g
      class="g-markers"
    />
    <g
      class="g-droplines"
    />
    <g
      class="g-tooltip-overlay"
    />
    <g
      class="g-legend-overlay"
    />
  </svg>
</DocumentFragment>
`;
